lib/verk/job.ex

```diff
@@ -11,18 +11,38 @@ defmodule Verk.Job do
   @type t :: %__MODULE__{error_message: String.t, failed_at: DateTime.t, retry_count: non_neg_integer,
                          queue: String.t, class: String.t | atom, jid: String.t, finished_at: DateTime.t,
                          retried_at: DateTime.t, error_backtrace: String.t}
-  @derive {Poison.Encoder, only: Keyword.keys(@keys)}
   defstruct [:original_json | @keys]
 
+  @doc """
+  Encode the struct to a JSON string, raising if there is an error
+  """
+  @spec encode!(t) :: binary
+  def encode!(job = %__MODULE__{}) do
+    job
+    |> Map.from_struct()
+    |> Map.take(Keyword.keys(@keys))
+    # We wrap the args so there is no issue with cJson on Redis side
+    # It's just an opaque string as far as Redis is concerned
+    |> Map.update!(:args, &Jason.encode!(&1))
+    |> Jason.encode!()
+  end
+
   @doc """
   Decode the JSON payload storing the original json as part of the struct.
   """
-  @spec decode(binary) :: {:ok, %__MODULE__{}} | {:error, Poison.Error.t}
+  @spec decode(binary) :: {:ok, %__MODULE__{}} | {:error, Jason.DecodeError.t()}
   def decode(payload) do
-    case Poison.decode(payload, as: %__MODULE__{}) do
-      {:ok, job} -> {:ok, build(job, payload)}
-      {:error, error} -> {:error, error}
-      {:error, :invalid, pos} -> {:error, "Invalid json at position: #{pos}"}
+    with {:ok, map} <- Jason.decode(payload),
+         {:ok, args} <- unwrap_args(map["args"]) do
+      fields =
+        map
+        |> Map.update!("args", fn _ -> args end)
+        |> Map.new(fn {k, v} -> {String.to_atom(k), v} end)
```

Comment:
Do we need to convert to atoms?

Reply:
Yes exactly, from the doc:
```
Keys in the Enumerable that don’t exist in the struct are automatically discarded.
Note that keys must be atoms, as only atoms are allowed when defining a struct.
```

Reply:
There is an option in `Jason.decode/2` that handles this automatically, `keys: :atoms`. You can also use `existing_atoms` to be safer, IIRC.

Reply:
Ha good to known @sorentwo, I will leverage this then

Reply:
I think you need to define a struct but not to create a new one:

```
iex(5)> struct(Verk.Job, jid: "123")
%Verk.Job{
  args: [],
  class: nil,
  enqueued_at: nil,
  error_backtrace: nil,
  error_message: nil,
  failed_at: nil,
  finished_at: nil,
  jid: "123",
  max_retry_count: 25,
  original_json: nil,
  queue: nil,
  retried_at: nil,
  retry_count: 0
}
iex(6)> struct(Verk.Job, "jid": "123")
%Verk.Job{
  args: [],
  class: nil,
  enqueued_at: nil,
  error_backtrace: nil,
  error_message: nil,
  failed_at: nil,
  finished_at: nil,
  jid: "123",
  max_retry_count: 25,
  original_json: nil,
  queue: nil,
  retried_at: nil,
  retry_count: 0
}
```

Reply:
Nice! I completely forgot you had to do this 👍